src/Receiver.cc

```diff
@@ -90,83 +108,91 @@ Receiver::handleDataPacket(Driver::Packet* packet, IpAddress sourceIp)
-    uint16_t dataHeaderLength = sizeof(Protocol::Packet::DataHeader);
     Protocol::MessageId id = header->common.messageId;
 
+    bool needSchedule = false;
+    bool messageComplete;
     MessageBucket* bucket = messageBuckets.getBucket(id);
-    SpinLock::Lock lock_bucket(bucket->mutex);
-    Message* message = bucket->findMessage(id, lock_bucket);
-    if (message == nullptr) {
-        // New message
-        int messageLength = header->totalLength;
-        int numUnscheduledPackets = header->unscheduledIndexLimit;
-        {
-            SpinLock::Lock lock_allocator(messageAllocator.mutex);
+    Message* message;
+    {
+        // Scoped critical section guarded by MessageBucket::mutex; this ensures
+        // that the bucket mutex is dropped before acquiring the schedulerMutex.
+        SpinLock::Lock lock_bucket(bucket->mutex);
+        message = bucket->findMessage(id, lock_bucket);
+        if (message == nullptr) {
+            // New message
+            int messageLength = header->totalLength;
+            int numUnscheduledPackets = header->unscheduledIndexLimit;
             SocketAddress srcAddress = {
                 .ip = sourceIp, .port = be16toh(header->common.prefix.sport)};
-            message = messageAllocator.pool.construct(
-                this, driver, dataHeaderLength, messageLength, id, srcAddress,
-                numUnscheduledPackets);
+            message =
+                messageAllocator.construct(this, driver, messageLength, id,
+                                           srcAddress, numUnscheduledPackets);
             Perf::counters.allocated_rx_messages.add(1);
-        }
 
-        bucket->messages.push_back(&message->bucketNode);
-        policyManager->signalNewMessage(
-            message->source.ip, header->policyVersion, header->totalLength);
+            // Start tracking the message.
+            bucket->messages.push_back(&message->bucketNode);
+            policyManager->signalNewMessage(
+                message->source.ip, header->policyVersion, header->totalLength);
 
-        if (message->scheduled) {
-            // Message needs to be scheduled.
-            SpinLock::Lock lock_scheduler(schedulerMutex);
-            schedule(message, lock_scheduler);
+            if (message->scheduled) {
+                // Don't schedule the message while holding the bucket mutex.
+                needSchedule = true;
+            }
         }
+
+        // Add the packet, but don't copy the payload yet.
+        if (message->occupied.test(header->index)) {
+            // Must be a duplicate packet; drop it.
+            return;
+        }
+        message->occupied.set(header->index);
+        message->numPackets++;
+        messageComplete = (message->numPackets == message->numExpectedPackets);
     }
 
-    // Things that must be true (sanity check)
-    assert(id == message->id);
-    assert(message->driver == driver);
+    // Copy the payload into the message buffer.
```

Comment:
Why copy the data out while processing the packet instead of when the application calls `receive()`? I can think of potential pros and cons but I'm wondering what you are considering.

Reply:
If you want to be able to release the packet buffer back to the driver when this method returns, you would have to do the copy here. In general, it's going to be faster to do copy while the packet data is still hot; this also helps reduce the number of RX buffers in use in the driver.